[CDC] NPE in ChangeRecordImpl#hashCode for INSERT/SYNC operations

**Describe the bug**

hashCode method in ChangeRecordImpl uses .toJson() method on oldValue/newValue fields but for SYNC/INSERT operations the oldValue is nullable.
Example on screenshot:
![image](https://user-images.githubusercontent.com/20545793/195076544-fc822d45-5c93-4430-aac2-fe37315c36e1.png)

Logs:
```
Caused by: java.lang.NullPointerException
	at com.hazelcast.jet.cdc.impl.ChangeRecordImpl.hashCode(ChangeRecordImpl.java:183) ~[classes/:?]
	at com.hazelcast.jet.core.Partitioner.lambda$static$7a6ef289$1(Partitioner.java:70) ~[classes/:?]
	at com.hazelcast.jet.core.Edge$KeyPartitioner.getPartition(Edge.java:818) ~[classes/:?]
	at com.hazelcast.jet.impl.execution.OutboundCollector$Partitioned.offer(OutboundCollector.java:198) ~[classes/:?]
	at com.hazelcast.jet.impl.execution.OutboxImpl.doOffer(OutboxImpl.java:255) ~[classes/:?]
	at com.hazelcast.jet.impl.execution.OutboxImpl.offerInternal(OutboxImpl.java:142) ~[classes/:?]
	at com.hazelcast.jet.impl.execution.OutboxImpl.offer(OutboxImpl.java:99) ~[classes/:?]
	at com.hazelcast.jet.core.AbstractProcessor.tryEmit(AbstractProcessor.java:309) ~[classes/:?]
	at com.hazelcast.jet.core.AbstractProcessor.emitFromTraverser(AbstractProcessor.java:408) ~[classes/:?]
	at com.hazelcast.jet.core.AbstractProcessor.emitFromTraverser(AbstractProcessor.java:421) ~[classes/:?]
	at com.hazelcast.jet.cdc.impl.CdcSourceP.complete(CdcSourceP.java:189) ~[classes/:?]
	at com.hazelcast.jet.impl.execution.ProcessorTasklet.complete(ProcessorTasklet.java:541) ~[classes/:?]
	at com.hazelcast.jet.impl.execution.ProcessorTasklet.stateMachineStep(ProcessorTasklet.java:421) ~[classes/:?]
	at com.hazelcast.jet.impl.execution.ProcessorTasklet.call(ProcessorTasklet.java:291) ~[classes/:?]
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$BlockingWorker.run(TaskletExecutionService.java:315) ~[classes/:?]

```
**Expected behavior**
hashCode method should be null-safe (for oldValue and newValue)

**To Reproduce**

Steps to reproduce the behavior:
1. Go to 'com.hazelcast.jet.cdc.DebeziumCdcIntegrationTest'
2. Replace current pipeline definition for mysyql method with:
```
           Pipeline pipeline = Pipeline.create();
            StreamSource<ChangeRecord> source = mySqlSource(container);
            pipeline.readFrom(source)
                    .withNativeTimestamps(1)
                    .setLocalParallelism(1)
                    .window(SlidingWindowDefinition.sliding(SECONDS.toMillis(2), SECONDS.toMillis(1)))
                    .distinct()
                    .writeTo(Sinks.logger(WindowResult::toString));
```
3. Above pipeline will throw NPE due to nullable "oldValue for SYNC entries"

**Additional context**
Found during testing on 5.2.0-SNAPSHOT

--- a/extensions/cdc-debezium/src/main/java/com/hazelcast/jet/cdc/impl/ChangeRecordImpl.java
+++ b/extensions/cdc-debezium/src/main/java/com/hazelcast/jet/cdc/impl/ChangeRecordImpl.java
@@ -177,7 +177,8 @@ public class ChangeRecordImpl implements ChangeRecord {
 
     @Override
     public int hashCode() {
-        return Objects.hash(sequenceSource, sequenceValue, keyJson, oldValue, newValue);
+        return Objects.hash(sequenceSource, sequenceValue, keyJson, timestamp,
+                operation, database, schema, table, oldValue, newValue);
     }
 
     @Override
@@ -192,6 +193,11 @@ public class ChangeRecordImpl implements ChangeRecord {
         return this.sequenceSource == that.sequenceSource
                 && this.sequenceValue == that.sequenceValue
                 && Objects.equals(this.keyJson, that.keyJson)
+                && Objects.equals(this.timestamp, that.timestamp)
+                && this.operation == that.operation
+                && Objects.equals(this.database, that.database)
+                && Objects.equals(this.schema, that.schema)
+                && Objects.equals(this.table, that.table)
                 && Objects.equals(this.oldValue, that.oldValue)
                 && Objects.equals(this.newValue, that.newValue);
     }

--- a/extensions/cdc-debezium/src/main/java/com/hazelcast/jet/cdc/impl/ChangeRecordImpl.java
+++ b/extensions/cdc-debezium/src/main/java/com/hazelcast/jet/cdc/impl/ChangeRecordImpl.java
@@ -177,12 +177,7 @@ public class ChangeRecordImpl implements ChangeRecord {
 
     @Override
     public int hashCode() {
-        int hash = (int) sequenceSource;
-        hash = 31 * hash + (int) sequenceValue;
-        hash = 31 * hash + keyJson.hashCode();
-        hash = 31 * hash + oldValue.toJson().hashCode();
-        hash = 31 * hash + newValue.toJson().hashCode();
-        return hash;
+        return Objects.hash(sequenceSource, sequenceValue, keyJson, oldValue, newValue);
     }
 
     @Override

--- a/extensions/cdc-debezium/src/main/java/com/hazelcast/jet/cdc/impl/ChangeRecordImpl.java
+++ b/extensions/cdc-debezium/src/main/java/com/hazelcast/jet/cdc/impl/ChangeRecordImpl.java
@@ -189,10 +189,10 @@ public class ChangeRecordImpl implements ChangeRecord {
             return false;
         }
         ChangeRecordImpl that = (ChangeRecordImpl) obj;
-        return this.sequenceSource == that.sequenceSource &&
-                this.sequenceValue == that.sequenceValue &&
-                this.keyJson.equals(that.keyJson) &&
-                Objects.equals(this.oldValue, that.oldValue) &&
-                Objects.equals(this.newValue, that.newValue);
+        return this.sequenceSource == that.sequenceSource
+                && this.sequenceValue == that.sequenceValue
+                && this.keyJson.equals(that.keyJson)
+                && Objects.equals(this.oldValue, that.oldValue)
+                && Objects.equals(this.newValue, that.newValue);
     }
 }

--- a/extensions/cdc-debezium/src/main/java/com/hazelcast/jet/cdc/impl/ChangeRecordImpl.java
+++ b/extensions/cdc-debezium/src/main/java/com/hazelcast/jet/cdc/impl/ChangeRecordImpl.java
@@ -191,7 +191,7 @@ public class ChangeRecordImpl implements ChangeRecord {
         ChangeRecordImpl that = (ChangeRecordImpl) obj;
         return this.sequenceSource == that.sequenceSource
                 && this.sequenceValue == that.sequenceValue
-                && this.keyJson.equals(that.keyJson)
+                && Objects.equals(this.keyJson, that.keyJson)
                 && Objects.equals(this.oldValue, that.oldValue)
                 && Objects.equals(this.newValue, that.newValue);
     }

--- a/extensions/cdc-debezium/src/test/java/com/hazelcast/jet/cdc/DebeziumCdcIntegrationTest.java
+++ b/extensions/cdc-debezium/src/test/java/com/hazelcast/jet/cdc/DebeziumCdcIntegrationTest.java
@@ -421,6 +421,9 @@ public class DebeziumCdcIntegrationTest extends AbstractCdcIntegrationTest {
                     .window(sliding(SECONDS.toMillis(2), SECONDS.toMillis(1)))
                     .distinct()
                     .writeTo(Sinks.logger(WindowResult::toString));
+
+            HazelcastInstance hz = createHazelcastInstances(1)[0];
+            hz.getJet().newJob(pipeline);
         }
     }
 

--- a/extensions/cdc-debezium/src/test/java/com/hazelcast/jet/cdc/DebeziumCdcIntegrationTest.java
+++ b/extensions/cdc-debezium/src/test/java/com/hazelcast/jet/cdc/DebeziumCdcIntegrationTest.java
@@ -22,6 +22,7 @@ import com.hazelcast.jet.JetException;
 import com.hazelcast.jet.Job;
 import com.hazelcast.jet.accumulator.LongAccumulator;
 import com.hazelcast.jet.core.JobStatus;
+import com.hazelcast.jet.datamodel.WindowResult;
 import com.hazelcast.jet.pipeline.Pipeline;
 import com.hazelcast.jet.pipeline.Sinks;
 import com.hazelcast.jet.pipeline.StreamSource;
@@ -45,6 +46,8 @@ import java.util.Map.Entry;
 import java.util.Objects;
 
 import static com.hazelcast.jet.Util.entry;
+import static com.hazelcast.jet.pipeline.WindowDefinition.sliding;
+import static java.util.concurrent.TimeUnit.SECONDS;
 import static org.assertj.core.api.Assertions.assertThatThrownBy;
 import static org.testcontainers.containers.MySQLContainer.MYSQL_PORT;
 import static org.testcontainers.containers.PostgreSQLContainer.POSTGRESQL_PORT;
@@ -402,6 +405,23 @@ public class DebeziumCdcIntegrationTest extends AbstractCdcIntegrationTest {
         assertThatThrownBy(job::join)
                 .hasRootCauseInstanceOf(JetException.class)
                 .hasStackTraceContaining("connector class io.debezium.connector.xxx.BlaBlaBla not found");
+    }
+
+    @Test
+    public void notFailWhenOldValueNotPresent() {
+        try (MySQLContainer<?> container = mySqlContainer()) {
+            container.start();
+            Pipeline pipeline = Pipeline.create();
+
+            // distinct causes hashCode to be used
+            StreamSource<ChangeRecord> source = mySqlSource(container);
+            pipeline.readFrom(source)
+                    .withNativeTimestamps(1)
+                    .setLocalParallelism(1)
+                    .window(sliding(SECONDS.toMillis(2), SECONDS.toMillis(1)))
+                    .distinct()
+                    .writeTo(Sinks.logger(WindowResult::toString));
+        }
     }
 
     private static class Customer {

--- a/extensions/cdc-debezium/src/test/java/com/hazelcast/jet/cdc/DebeziumCdcIntegrationTest.java
+++ b/extensions/cdc-debezium/src/test/java/com/hazelcast/jet/cdc/DebeziumCdcIntegrationTest.java
@@ -22,7 +22,6 @@ import com.hazelcast.jet.JetException;
 import com.hazelcast.jet.Job;
 import com.hazelcast.jet.accumulator.LongAccumulator;
 import com.hazelcast.jet.core.JobStatus;
-import com.hazelcast.jet.datamodel.WindowResult;
 import com.hazelcast.jet.pipeline.Pipeline;
 import com.hazelcast.jet.pipeline.Sinks;
 import com.hazelcast.jet.pipeline.StreamSource;
@@ -46,8 +45,7 @@ import java.util.Map.Entry;
 import java.util.Objects;
 
 import static com.hazelcast.jet.Util.entry;
-import static com.hazelcast.jet.pipeline.WindowDefinition.sliding;
-import static java.util.concurrent.TimeUnit.SECONDS;
+import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.api.Assertions.assertThatThrownBy;
 import static org.testcontainers.containers.MySQLContainer.MYSQL_PORT;
 import static org.testcontainers.containers.PostgreSQLContainer.POSTGRESQL_PORT;
@@ -413,17 +411,26 @@ public class DebeziumCdcIntegrationTest extends AbstractCdcIntegrationTest {
             container.start();
             Pipeline pipeline = Pipeline.create();
 
-            // distinct causes hashCode to be used
+            // stateful transform causes hashCode to be used
             StreamSource<ChangeRecord> source = mySqlSource(container);
             pipeline.readFrom(source)
                     .withNativeTimestamps(1)
                     .setLocalParallelism(1)
-                    .window(sliding(SECONDS.toMillis(2), SECONDS.toMillis(1)))
-                    .distinct()
-                    .writeTo(Sinks.logger(WindowResult::toString));
+                    .groupingKey(r -> r)
+                    .mapStateful(
+                            LongAccumulator::new,
+                            (acc, key, record) -> {
+                                acc.add(1);
+                                return acc.get();
+                            }
+                    )
+                    .peek()
+                    .writeTo(Sinks.list("notFailWhenOldValueNotPresent"));
 
             HazelcastInstance hz = createHazelcastInstances(1)[0];
             hz.getJet().newJob(pipeline);
+
+            assertTrueEventually(() -> assertThat(hz.getList("notFailWhenOldValueNotPresent")).isNotEmpty());
         }
     }
 

